### PR TITLE
Updated OnUnitDied to properly kill gremlins when their owner dies

### LINKED ISSUE #367: Gremlins don't die when their owner dies under certain circumstances
https://github.com/X2CommunityCore/X2WOTCCommunityHighlander/blob/50d24a1fe39bf32492ff28f82296dfb91b4adaaa/X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Item.uc#L477-L502

    //  if it was not our owner, reattach to them 
    if ( ( UnitState.ObjectID != OwnerStateObject.ObjectID ) || (OwnerStateObject.ObjectID != AttachedUnitRef.ObjectID) )

Bad. When Aid Protocol is active, and the gremlin's owner dies:

* `UnitState` is the unit that died (owner)
* `OwnerStateObject` is the owner
* `AttachedUnitRef` is the unit we are attached to

Thus, the condition evaluates to `(false || true)`, and the gremlin doesn't die. This causes major issues with enemy gremlins, as the sweep objective won't be able to complete (unless using `KillAllAIs`).

------------

I meant to report this... *checks Discord*... fifteen months ago, but I seem to have forgotten about it. Whoops.

### PR DESCRIPTION
**Pull request by BlackDog86**
Fixes #367

Tested under 4 standard conditions:

Gremlin attached to specialist 1 - Kill specialist 1 - Gremlin dies
Gremlin attached to another unit - Kill other unit - Gremlin returns to specialist 1
Gremlin attached to another unit - Kill specialist 1 - Gremlin dies
Gremlin attached to specialist 1 - Kill other unit - No effect

And 4 non-standard conditions:

Specialist 1 uses aid protocol on specialist 2 - Kill specialist 2 - Gremlin 1 returns to Specialist 1, Gremlin 2 dies
Specialist 1 uses aid protocol on specialist 2 - Kill specialist 1 - Gremlin 1 dies, Gremlin 2 no effect
Specialist 1 & 2 use aid protocol on each other - Kill both specialists at once - Both gremlins die
Specialist 1 & 2 use aid protocol on 2 different units - Kill both units at once - Both gremlins return

### CODE REVIEW

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Item.uc [hunk 1]
```diff
@@ -473,36 +473,31 @@ private function RecallItem(const XComGameState_Item RecalledItem, XComGameState
...
 	local vector NewLocation;
 	local XComGameStateContext_ChangeContainer ChangeContext;
-
+	
+	// Start Issue #367
+	/// HL-Docs: ref:Bugfixes; issue:367
+	/// Structure of the OnUnitDied function is adjusted to respect the case when the gremlin's owner is killed, 
+	/// but the gremlin its-self is attached to another unit. The adjusted code now does not overwrite the unitstate
+	/// of the unit that died with the item owner unit-state halfway through the function, making the code path 
+	/// easier to follow.
```
Comment (robojumper) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Item.uc:488]:
```suggestion
	/// Gremlins now correctly die when the Gremlin's owner dies while the Gremlin is attached to another unit.
```

I indented for Bugfix documentation to be shorter and worded in a user-facing way, so maybe this is enough? C.f. https://x2communitycore.github.io/X2WOTCCommunityHighlander/Bugfixes/

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Item.uc [hunk 2]
```diff
@@ -473,36 +473,31 @@ private function RecallItem(const XComGameState_Item RecalledItem, XComGameState
...
-				XGUnit(CosmeticUnit.GetVisualizer()).MoveToLocation(NewLocation);
-			}
-		}
-		//  if it was our owner, we have to die too
-		else
-		{
+		
+		// If the unit that died was the owner of the gremlin, kill the (poor little) gremlin too
+		If(UnitState.ObjectID == OwnerUnitState.ObjectID)
+		{
```
Comment (robojumper) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Item.uc:500]: Can we change this condition to `!=` and swap the branches to more closely keep the original code structure?

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Item.uc (hunk 1 above)
Comment (BlackDog86) [reply X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Item.uc]: Done :)

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Item.uc (hunk 2 above)
Comment (BlackDog86) [reply X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Item.uc]: And done :)

**Review by robojumper — COMMENTED** [review]
This could use a minor formatting pass - the original code doesn't look great either but this *adds* trailing whitespace. Also `If(UnitState.ObjectID != OwnerUnitState.ObjectID)` -> `if (UnitState.ObjectID != OwnerUnitState.ObjectID)`

**Comment by BlackDog86** [comment]
Should be sorted

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Item.uc [hunk 3]
```diff
@@ -473,34 +473,40 @@ private function RecallItem(const XComGameState_Item RecalledItem, XComGameState
...
 	local XComGameState_Item ItemState;
 	local vector NewLocation;
 	local XComGameStateContext_ChangeContainer ChangeContext;
-
+	
+	// Start Issue #367
+	/// HL-Docs: ref:Bugfixes; issue:367
+	/// Gremlins now correctly die when the Gremlin's owner dies while the Gremlin is attached to another unit.
+	
 	UnitState = XComGameState_Unit(EventData);
```
Comment (Iridar) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Item.uc:487]: The original code is a bit awful about none-checking the unit state objects. Ideally, the checks should be done either right after getting the objects, or right before using them. This code does `if (UnitState.ObjectID != OwnerUnitState.ObjectID)`, and only then `if (OwnerUnitState != none`. `UnitState` itself is no longer none-checked at all. 

Obviously these are edge cases, but my experience tells me it's better to be extra careful with this.

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Item.uc [hunk 4]
```diff
@@ -473,34 +473,40 @@ private function RecallItem(const XComGameState_Item RecalledItem, XComGameState
...
-	//  was this the unit we are attached to or our owner?
+	// Check if the unit that died had the gremlin attached to it, or is the owner of the gremlin
 	if (UnitState.ObjectID == AttachedUnitRef.ObjectID || UnitState.ObjectID == OwnerStateObject.ObjectID)
```
Comment (Iridar) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Item.uc:489]: With above in mind, I would change `if (UnitState.ObjectID == AttachedUnitRef.ObjectID || UnitState.ObjectID == OwnerStateObject.ObjectID)` into `if (UnitState != none && (UnitState.ObjectID == AttachedUnitRef.ObjectID || UnitState.ObjectID == OwnerStateObject.ObjectID))`

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Item.uc [hunk 5]
```diff
@@ -473,34 +473,40 @@ private function RecallItem(const XComGameState_Item RecalledItem, XComGameState
...
 	{
 		History = `XCOMHISTORY;
-		UnitState = XComGameState_Unit(History.GetGameStateForObjectID(OwnerStateObject.ObjectID));
+		OwnerUnitState = XComGameState_Unit(History.GetGameStateForObjectID(OwnerStateObject.ObjectID));
 		CosmeticUnit = XComGameState_Unit(History.GetGameStateForObjectID(CosmeticUnitRef.ObjectID));
-		//  if it was not our owner, reattach to them
-		if ( ( UnitState.ObjectID != OwnerStateObject.ObjectID ) || (OwnerStateObject.ObjectID != AttachedUnitRef.ObjectID) )
+		
+		// If the unit that died was not the owner of the gremlin, recall it & re-attach it to its original owner
+		if (UnitState.ObjectID != OwnerUnitState.ObjectID)
```
Comment (Iridar) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Item.uc:496]: And this into: `if (OwnerUnitState != none && UnitState.ObjectID != OwnerUnitState.ObjectID)`

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Item.uc [hunk 6]
```diff
@@ -473,34 +473,40 @@ private function RecallItem(const XComGameState_Item RecalledItem, XComGameState
...
 		{
 			NewGameState = class'XComGameStateContext_ChangeContainer'.static.CreateChangeState("Attached Unit Died");
 			ItemState = XComGameState_Item(NewGameState.ModifyStateObject(Class, ObjectID));
 			ItemState.AttachedUnitRef = OwnerStateObject;
 			`GAMERULES.SubmitGameState(NewGameState);
-						
-			if (UnitState != none && CosmeticUnit != none && CosmeticUnit.TileLocation != UnitState.TileLocation)
+
+			if (OwnerUnitState != none && CosmeticUnit != none && CosmeticUnit.TileLocation != OwnerUnitState.TileLocation)
```
Comment (Iridar) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Item.uc:503]: And then `OwnerUnitState != none` can be removed from here.

**Review by Iridar — CHANGES_REQUESTED** [review]
Just make sure the codepath is not altered by none-checks. I.e. code that would run if either of the state objects was `none` should still run, as long as it doesn't actually require said state objects.

**Comment by Iridar** [comment]
I've refactored the code, could you please redo your four tests with this code?
```cpp
function EventListenerReturn OnUnitDied(Object EventData, Object EventSource, XComGameState GameState, Name Event, Object CallbackData)
{
	local XComGameStateHistory History;
	local XComGameState_Unit UnitState, CosmeticUnit;
	local XComGameState NewGameState;
	local XComGameState_Item ItemState;
	local vector NewLocation;
	local XComGameStateContext_ChangeContainer ChangeContext;
	
	// Variable for Issue #367
	local XComGameState_Unit OwnerUnitState;
	
	// Start Issue #367
	/// HL-Docs: ref:Bugfixes; issue:367
	/// Gremlins now correctly die when the Gremlin's owner dies while the Gremlin is attached to another unit.
	
	UnitState = XComGameState_Unit(EventData);
	if (UnitState == none)
		return ELR_NoInterrupt;
	
	if (UnitState.ObjectID == OwnerStateObject.ObjectID)
	{
		// Owner of the Gremlin died, kill the Gremlin.
		
		History = `XCOMHISTORY;
		CosmeticUnit = XComGameState_Unit(History.GetGameStateForObjectID(CosmeticUnitRef.ObjectID));
		
		if (CosmeticUnit != none)
		{
			NewGameState = class'XComGameStateContext_ChangeContainer'.static.CreateChangeState("Owner Unit Died");
			ChangeContext = XComGameStateContext_ChangeContainer(NewGameState.GetContext());
			ChangeContext.BuildVisualizationFn = ItemOwnerDeathVisualization;
			CosmeticUnit = XComGameState_Unit(NewGameState.ModifyStateObject(CosmeticUnit.Class, CosmeticUnit.ObjectID));
			CosmeticUnit.SetCurrentStat(eStat_HP, 0);
			`GAMERULES.SubmitGameState(NewGameState);
		}
	}
	else if (UnitState.ObjectID == AttachedUnitRef.ObjectID)
	{
		// Unit to whom the Gremlin is attached to has died, recall the Gremlin and re-attach to its owner.
		
		History = `XCOMHISTORY;
		CosmeticUnit = XComGameState_Unit(History.GetGameStateForObjectID(CosmeticUnitRef.ObjectID));
		OwnerUnitState = XComGameState_Unit(History.GetGameStateForObjectID(OwnerStateObject.ObjectID));
		
		if (OwnerUnitState != none && CosmeticUnit != none)
		{
			NewGameState = class'XComGameStateContext_ChangeContainer'.static.CreateChangeState("Attached Unit Died");
			ItemState = XComGameState_Item(NewGameState.ModifyStateObject(Class, ObjectID));
			ItemState.AttachedUnitRef = OwnerStateObject;
			`GAMERULES.SubmitGameState(NewGameState);

			if (CosmeticUnit.TileLocation != OwnerUnitState.TileLocation)
			{
				NewLocation = `XWORLD.GetPositionFromTileCoordinates(OwnerUnitState.TileLocation);
				XGUnit(CosmeticUnit.GetVisualizer()).MoveToLocation(NewLocation);
			}
		}
	}
	// End Issue #367
	
	return ELR_NoInterrupt;
}
```

The bug with the original code happened mostly because it was convoluted and hard to read. Bad code. I don't doubt that the code you submitted fixes the bug, but it's still hard to read, so I've simplified it as much as possible.

**Comment by BlackDog86** [comment]
Refactored code testing:

Gremlin attached to specialist 1 - Kill specialist 1 - Gremlin dies CONFIRMED
Gremlin attached to another unit - Kill other unit - Gremlin returns to specialist 1 CONFIRMED
Gremlin attached to another unit - Kill specialist 1 - Gremlin dies CONFIRMED
Gremlin attached to specialist 1 - Kill other unit - No effect CONFIRMED

And 4 non-standard conditions:

Specialist 1 uses aid protocol on specialist 2 - Kill specialist 2 - Gremlin 1 returns to Specialist 1, Gremlin 2 dies CONFIRMED
Specialist 1 uses aid protocol on specialist 2 - Kill specialist 1 - Gremlin 1 dies, Gremlin 2 no effect CONFIRMED
Specialist 1 & 2 use aid protocol on each other - Kill both specialists at once - Both gremlins die CONFIRMED
Specialist 1 & 2 use aid protocol on 2 different units - Kill both units at once - Both gremlins return CONFIRMED

Looks like everything is fine with the refactor

**Comment by Iridar** [comment]
Thank you.

**Pull request merged**